swarm/swarm_test.go

```diff
@@ -17,11 +17,165 @@
 package swarm
 
 import (
+	"io/ioutil"
+	"path"
 	"testing"
 
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm/api"
 )
 
+// TestNewSwarm validates Swarm fields in repsect to the provided configuration.
+func TestNewSwarm(t *testing.T) {
+	dir, err := ioutil.TempDir("", "swarm")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// a simple rpc endpoint for testing dialing
+	ipcEndpoint := path.Join(dir, "TestSwarm.ipc")
+
+	_, server, err := rpc.StartIPCEndpoint(ipcEndpoint, nil)
+	if err != nil {
+		t.Error(err)
+	}
+	defer server.Stop()
+
+	for _, tc := range []struct {
+		name      string
+		configure func(*api.Config)
+		check     func(*testing.T, *Swarm, *api.Config)
+	}{
+		{
+			name:      "defaults",
+			configure: nil,
+			check: func(t *testing.T, s *Swarm, config *api.Config) {
+				if s.config != config {
+					t.Error("config is not the same object")
+				}
+				if s.backend != nil {
+					t.Error("backend is not nil")
+				}
+				if s.privateKey == nil {
+					t.Error("private key is not set")
+				}
+				if !s.config.HiveParams.Discovery {
+					t.Error("config.HiveParams.Discovery is false, must be true regardless the configuration")
```

Comment:
Why do we have configuration parameters if it always has to be true?

Reply:
Exactly my question in this PR https://github.com/ethersphere/go-ethereum/pull/557#discussion-diff-189232763L125. That is the case that we have now.